### ISSUE TITLE
Screenreader Feature Request for new Responsive Web Design curriculum

### ISSUE THREAD

**Issue by QuincyLarson** [freeCodeCamp/freeCodeCamp#45419]
I received this feedback from a camper who codes using the assistance of a screen reader.

Since we take accessibility seriously, I wanted to report their feedback verbatim. I've asked them to comment on this issue so that they can give feedback on our solution to this issue.

"""
I have an issue. I'm visually impaired. and facing an issue with accessibility. in the new responsive web design beta certification, I'm trying to find the remaining steps in my current course. But there is no any indication with screen reader.

In the old certification, on every step, there was an indication passed, or not passed. Even before starting the course, there was a summary of the course, for example, Not Passed 6/44.

But in the latest one, this is a problem. I can’t get my progress if I’ve forgot.

And the other problem, in the new certification, when you expand or collapse a course, there is no indication with screen reader like in the old one.
"""

**Comment by arqamgrt** [comment]
Hi there. Yes, I've reported this issue to @QuincyLarson. Thanks for creating this issue.
few days ago, it was fine, but yesterday, I tried and faced this issue.

**Comment by moT01** [comment]
Thanks for bringing this to our attention @argamgrt. So it seems like most of the problems are on the beta responsive web design map page, is that correct?

I took a look at the new and old Responsive Web Design (RWD) using the screen reader on my Mac. The tables below describe my experience. The items on the left were the order in which the items were selected. Anything in quotations are examples of what my screen reader read or what I think it should read.

### Old RWD:
| Item                    | Current Behavior                                                                                                                       | Issues                                                                                                                                                       | Potential Desired Behavior                                   |
|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------|
| Block Title             | "Heading level 3, Basic HTML and HTML5"                                                                                                |                                                                                                                                                                      | "Heading level 3, Basic HTML and HTML5, 1 of 28 passed"      |
| Expand, Collapse Button | "Collapse course, Not Passed, 1 slash 28, expanded, button" Second Example: "Expand course, Not Passed, 1 slash 28, collapsed, button" | It says "slash", and "passed" or "not passed". We could combine those to say "1 of 28 passed". We could perhaps move the overall progress to the block heading above | "Collapse course, expanded, button"                          |
| List of Challenges      | "List, 28 items"                                                                                                                       |                                                                                                                                                                      | "List, 28 items"                                             |
| Single Challenge        | "Visited, Link, Passed, Say Hello to HTML Elements, 1 of 28" Second example: "Link, Not Passed, Headline with the h2 element, 2 of 28" |                                                                                                                                                                      | "Visited, Link, Passed, Say Hello to HTML Elements, 1 of 28" |

### New (beta) RWD:
| Item                    | Current Behavior                                                                             | Issues                                                                                                                                                                                                                                                                                       | Potential Desired Behavior   |
|-------------------------|----------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------|
| Block Title             | "Visited, link, not passed, Learn HTML by Building a Cat Photo App, then all the intro text" | It's quite an earful to have all of this here, we could maybe split it up a little. It doesn't give you an overall progress level, e.g. "1 of 67 passed". It tells you it's a link when it's a button. It doesn't explain that it expands or collapses the map and that it doesn't take you anywhere | Something similar to old RWD                            |
| Expand, Collapse Button | This is combined into the block title on beta                                                |                                                                                                                                                                                                                                                                                                      |                             |
| List of Challenges      | "List, 67 items"                                                                             |                                                                                                                                                                                                                                                                                                      | "List, 67 items"             |
| Single Challenge        | "Link, 1, 1 of 67"                                                                           | It doesn't say whether it's passed or not. We probably don't need to read the text here. It could maybe say "Link, Passed, Step 1 of 67"                                                                                                                                                             | "Link, Passed, Step 1 of 67" |

I split it into two table because it got a little cluttered. I tried to give everyone an idea of the current behavior, what some of the issues are, and some potential solutions. I'm curious if you have any thoughts on this @arqamgrt. Is my experience similar to yours? and how would you like to see it behave?

**Comment by arqamgrt** [comment]
I completely agree with your detailed investigation. @moT01  :) I would like to see the solutions you've shared. this is what I want.

**Comment by bbsmooth** [comment]
I'm going to start working on beta accessibility as well. There are some easy wins, such as making the course toggle a disclosure button instead of a link (which addresses the problem of not being able to tell if the course expanded). And there there are more systemic issues like how do we tell users which steps they've completed and how do we get the user to their next uncompleted step without having to Tab through all the previous steps.

**Comment by superbkundan** [comment]
yes, it is very frustrating! I've just completed the first CSS course after the initial html on building the CatPhoto app, But, after completing, it says that I've completed it 98%. Now I've no way to know which part of the steps I've missed or not completed. Please do fix it as soon as possible.

**Comment by bbsmooth** [comment]
@superbkundan I was thinking of adding a shortcut link before the list of steps which allows the user to jump directly to their next uncompleted step. In your case, this would allow you to jump directly to the first missing step. In general, this would allow keyboard users to jump to the next uncompleted step without having to tab through all the previous completed steps. What do you think about this? Do you have any other suggestions to make navigating steps easier?

**Comment by arqamgrt** [comment]
@bbsmooth I agree with your suggestion. at least it will help screen reader users to directly jump to their incompleted tasks. You can work on this idea.
but at the same time. accessibility is still important.

**Comment by bbsmooth** [comment]
I have submitted my first PR for this issue (#45544). This PR is narrowly focused on just the toggle button heading for each course. It addresses two of the original issues:
- Notification that the course toggle has been expanded/collapsed
- Provide a course progress summary (i.e. how many steps have been completed).

**Comment by bbsmooth** [comment]
My first PR (#45544) related to this issue has been merged. Now I'll focus on the individual steps in the course.

**Comment by superbkundan** [comment]
No, How can this be closed without solving the issues? Did you add the
passed/not passed status in each of the steps of every curriculum? I just
checked now. You didn't. Plese do it.

**Comment by bbsmooth** [comment]
Hi @superbkundan I understand the confusion. It is closed because this fix has been merged into the main code base. But the new code hasn't been pushed to production yet. I don't know how often they push to production, but it should be showing up within the next few days I would think.

**Comment by superbkundan** [comment]
Ok, thanks for the update.

**Comment by superbkundan** [comment]
Great, thanks! Go on!
